SwiftCompilerSources/Sources/Optimizer/Utilities/BorrowUtils.swift

```diff
@@ -0,0 +1,516 @@
+    self.visitedReborrows = ValueSet(context)
+  }
+
+  mutating func deinitialize() {
+    visitedReborrows.deinitialize()
+  }
+
+  mutating func gather(enclosingValues: inout Stack<Value>,
+    forValue value: Value, _ cache: inout BorrowIntroducers.Cache) {
+    if value is Undef || value.ownership != .guaranteed {
+      return
+    }
+    var cache = BorrowIntroducers.Cache()
+    if BeginBorrowValue(value) != nil {
+      switch value {
+      case let beginBorrow as BeginBorrowInst:
+        // Gather the outer enclosing borrow scope.
+        BorrowIntroducers.gather(introducers: &enclosingValues,
+          forValue: beginBorrow.operand.value, &cache, context)
+      case is LoadBorrowInst, is FunctionArgument:
+        // There is no enclosing value on this path.
+        break
+      default:
+        gather(enclosingValues: &enclosingValues, forReborrow: Phi(value)!,
+          &cache)
+      }
+    } else {
+      // Handle forwarded guaranteed values.
+      BorrowIntroducers.gather(introducers: &enclosingValues, forValue: value,
+        &cache, context)
+    }
+  }
+  
+  // Given a reborrow, find the enclosing values. Each enclosing value
+  // is represented by one of the following cases, which refer to the
+  // example below:
+  //
+  // dominating owned value -> %value encloses %reborrow_1
+  // owned outer-adjacent phi -> %phi_3 encloses %reborrow_3
+  // dominating outer borrow introducer -> %outerBorrowB encloses %reborrow
+  // outer-adjacent reborrow -> %outerReborrow encloses %reborrow
+  //
+  // Recurse along the use-def phi web until a begin_borrow is
+  // reached. Then find all introducers of the begin_borrow's
+  // operand. At each level, find the outer adjacent phi, if one
+  // exists, otherwise return the most recently found dominating
+  // definition.
+  //
+  // If `reborrow` was already encountered because of a phi cycle,
+  // then no enclosingDefs are added.
+  //
+  // Example:
+  //
+  //         %value = ...
+  //         %borrow = begin_borrow %value
+  //         br one(%borrow)
+  //     one(%reborrow_1 : @guaranteed)
+  //         br two(%value, %reborrow_1)
+  //     two(%phi_2 : @owned, %reborrow_2 : @guaranteed)
+  //         br three(%value, %reborrow_1)
+  //     three(%phi_3 : @owned, %reborrow_3 : @guaranteed)
+  //         end_borrow %reborrow_3
+  //         destroy_value %phi_3
+  //
+  // gather(%reborrow_3) finds %phi_3 by computing enclosing defs in this order
+  // (inner -> outer):
+  //
+  //     %reborrow_1 -> %value
+  //     %reborrow_2 -> %phi_2
+  //     %reborrow_3 -> %phi_3
+  //
+  // Example:
+  //
+  //         %outerBorrowA = begin_borrow
+  //         %outerBorrowB = begin_borrow
+  //         %struct = struct (%outerBorrowA, outerBorrowB)
+  //         %borrow = begin_borrow %struct
+  //         br one(%outerBorrowA, %borrow)
+  //     one(%outerReborrow : @guaranteed, %reborrow : @guaranteed)
+  //
+  // gather(%reborrow) finds (%outerReborrow, %outerBorrowB).
```

Comment:
```
  // Example:
  //
  //         %outerBorrowA = begin_borrow
  //         %outerBorrowB = begin_borrow
  //         %struct = struct (%outerBorrowA, outerBorrowB)
  //         %borrow = begin_borrow %struct
  //         br one(%outerBorrowA, %borrow)
  //     one(%outerReborrow : @guaranteed, %reborrow : @guaranteed)
  //
  // gather(%reborrow) finds (%outerReborrow, %outerBorrowB).
```

There are only two enclosing defs of `%reborrow` : `%outerReborrow, %outerBorrowB`.

We only need to know that `%reborrow` keeps `%outerReborrow` alive. `%outerBorrowA` is an incoming value that will be kept alive by `%outerReborrow`, but in general, it might not even dominate `%reborrow`